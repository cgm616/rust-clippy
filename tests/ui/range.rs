#[warn(clippy::range_zip_with_len)]
fn main() {
    let v1 = vec![1, 2, 3];
    let v2 = vec![4, 5];
    let _x = v1.iter().zip(0..v1.len());
    let _y = v1.iter().zip(0..v2.len()); // No error
}

#[allow(unused)]
fn no_panic_with_fake_range_types() {
    struct Range {
        foo: i32,
    }

    let _ = Range { foo: 0 };
}

#[warn(clippy::manual_range_contains)]
#[allow(unused)]
#[allow(clippy::no_effect)]
#[allow(clippy::short_circuit_statement)]
fn manual_range_contains(x: u32) {
    // order shouldn't matter
    x >= 8 && x < 12;
    x < 42 && x >= 21;
    100 > x && 1 <= x;

    // also with inclusive ranges
    x >= 9 && x <= 99;
    x <= 33 && x >= 1;
    999 >= x && 1 <= x;

    // and the outside
    x < 8 || x >= 12;
    x >= 42 || x < 21;
    100 <= x || 1 > x;

    // also with the outside of inclusive ranges
    x < 9 || x > 99;
    x > 33 || x < 1;
    999 < x || 1 > x;

    // not a range.contains
    x > 8 && x < 12; // lower bound not inclusive
    x < 8 && x <= 12; // same direction
    x >= 12 && 12 >= x; // same bounds
    x < 8 && x > 12; // wrong direction

    x <= 8 || x >= 12;
    x >= 8 || x >= 12;
    x < 12 || 12 < x;
    x >= 8 || x <= 12;
}
