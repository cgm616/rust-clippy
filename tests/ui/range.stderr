error: it is more idiomatic to use `v1.iter().enumerate()`
  --> $DIR/range.rs:5:14
   |
LL |     let _x = v1.iter().zip(0..v1.len());
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::range-zip-with-len` implied by `-D warnings`

error: manual `Range::contains` implementation
  --> $DIR/range.rs:24:5
   |
LL |     x >= 8 && x < 12;
   |     ^^^^^^^^^^^^^^^^ help: use: `(8..12).contains(x)`
   |
   = note: `-D clippy::manual-range-contains` implied by `-D warnings`

error: manual `Range::contains` implementation
  --> $DIR/range.rs:25:5
   |
LL |     x < 42 && x >= 21;
   |     ^^^^^^^^^^^^^^^^^ help: use: `(21..42).contains(x)`

error: manual `Range::contains` implementation
  --> $DIR/range.rs:26:5
   |
LL |     100 > x && 1 <= x;
   |     ^^^^^^^^^^^^^^^^^ help: use: `(1..100).contains(x)`

error: manual `RangeInclusive::contains` implementation
  --> $DIR/range.rs:29:5
   |
LL |     x >= 9 && x <= 99;
   |     ^^^^^^^^^^^^^^^^^ help: use: `(9..=99).contains(x)`

error: manual `RangeInclusive::contains` implementation
  --> $DIR/range.rs:30:5
   |
LL |     x <= 33 && x >= 1;
   |     ^^^^^^^^^^^^^^^^^ help: use: `(1..=33).contains(x)`

error: manual `RangeInclusive::contains` implementation
  --> $DIR/range.rs:31:5
   |
LL |     999 >= x && 1 <= x;
   |     ^^^^^^^^^^^^^^^^^^ help: use: `(1..=999).contains(x)`

error: manual `!Range::contains` implementation
  --> $DIR/range.rs:34:5
   |
LL |     x < 8 || x >= 12;
   |     ^^^^^^^^^^^^^^^^ help: use: `!(8..12).contains(x)`

error: manual `!Range::contains` implementation
  --> $DIR/range.rs:35:5
   |
LL |     x >= 42 || x < 21;
   |     ^^^^^^^^^^^^^^^^^ help: use: `!(21..42).contains(x)`

error: manual `!Range::contains` implementation
  --> $DIR/range.rs:36:5
   |
LL |     100 <= x || 1 > x;
   |     ^^^^^^^^^^^^^^^^^ help: use: `!(1..100).contains(x)`

error: manual `!RangeInclusive::contains` implementation
  --> $DIR/range.rs:39:5
   |
LL |     x < 9 || x > 99;
   |     ^^^^^^^^^^^^^^^ help: use: `!(9..=99).contains(x)`

error: manual `!RangeInclusive::contains` implementation
  --> $DIR/range.rs:40:5
   |
LL |     x > 33 || x < 1;
   |     ^^^^^^^^^^^^^^^ help: use: `!(1..=33).contains(x)`

error: manual `!RangeInclusive::contains` implementation
  --> $DIR/range.rs:41:5
   |
LL |     999 < x || 1 > x;
   |     ^^^^^^^^^^^^^^^^ help: use: `!(1..=999).contains(x)`

error: aborting due to 13 previous errors

